Add blinking effect to signal ending of knight buff

ntwbruce opened the pull request:
Modify ExpireBuff such that when time left before buff ends is equal to a given blinking duration, it switches the knight's colour in between the default and the buffed colours repeatedly to simulate blinking, the length of each blink being a given period. Knight prefab is modified to set these values.

Currently the blinking is hard coded to 1.0s with a period of 0.2s, so 5 blinks.

https://user-images.githubusercontent.com/77188160/159859937-c2be0342-7106-4886-b3d7-62373d286381.mov

ntwbruce committed:
Add blinking effect to signal ending of knight buff

Abstract out methods for changing the knight's color. Add variables for
the total length of the blinking and time per blink. Once the buff has
time left < blinkingDuration, while loop repeatedly calls the methods to
remove and reapply the buff color, until end of blinkingDuration.

ntwbruce committed:
Update blinking period

ntwbruce committed:
Refactor to reduce RPC calls

Create separate subroutine to handle blinking, so there isn't one RPC
call for every blink/unblink.

Also, edit Knight prefab to preset blinkingDuration and blinkingPeriod.

lowkangn commented on Assets/Scripts/Actor Components/Buff.cs:
```diff
@@ -44,14 +49,36 @@ public void RemoveBuff()
             if (buffTimeout != null)
             {
                 StopCoroutine(buffTimeout);
+                StopCoroutine(blinkingTime);
```
There's the possibility of a null exception here if the Knight uses the buff before it starts blinking. I think you can simply reuse the same coroutine variable for both the bufftimeout and blinking coroutines. Alternatively you can also include checks for the blinkingtime coroutine as well but this might make the code harder to maintain.

lowkangn commented on Assets/Scripts/Actor Components/Buff.cs:
```diff
@@ -44,14 +49,36 @@ public void RemoveBuff()
+
+    private IEnumerator ExpireBlinking()
+    {
+        int numOfBlinks = (int) Math.Floor(Math.Min(blinkingDuration, buffDuration) / blinkingPeriod);
```
Good job on handling the case where the buff duration is too short/blinking duration is too long.

lowkangn commented on Assets/Scripts/Actor Components/Buff.cs:
```diff
@@ -74,4 +101,10 @@ private void RPC_RemoveBuff()
+    [PunRPC]
+    private void RPC_StartBlinking()
+    {
+        blinkingTime = StartCoroutine(ExpireBlinking());
```
As stated before, you can just reuse the blinktimeout variable.

lowkangn reviewed (commented):
Looks mostly good, but there are a few issues that require fixing.

ntwbruce committed:
Make minor fixes

- Revert changes to knight prefab, keep blinkingDuration and
blinkingPeriod to Buff script
- Use the same coroutine variable to avoid null exception

ntwbruce committed:
Remove unused variable

kenneth-fung reviewed (approved):
Looks good to me.
